engines/nancy/action/actionmanager.cpp

```diff
@@ -0,0 +1,466 @@
+#include "engines/nancy/ui/viewport.h"
+
+#include "common/memstream.h"
+#include "common/events.h"
+#include "common/str.h"
+#include "common/serializer.h"
+
+namespace Nancy {
+namespace Action {
+
+void ActionManager::handleInput(NancyInput &input) {
+    for (auto &rec : _records) {
+        if (rec->isActive) {
+            // Send input to all active records
+            rec->handleInput(input);
+        }
+
+        if (rec->isActive && rec->hasHotspot && NancySceneState.getViewport().convertViewportToScreen(rec->hotspot).contains(input.mousePos)) {
+            g_nancy->cursorManager->setCursorType(rec->getHoverCursor());
+
+            if (input.input & NancyInput::kLeftMouseButtonUp) {
+                input.input &= ~NancyInput::kLeftMouseButtonUp;
+
+                bool shouldTrigger = false;
+                int16 heldItem = NancySceneState.getHeldItem();
+                if (rec->itemRequired != -1) {
+                    if (heldItem == -1 && rec->itemRequired == -2) {
+                        shouldTrigger = true;
+                    } else {
+                        if (rec->itemRequired <= 100) {
+                            if (heldItem == rec->itemRequired) {
+                                shouldTrigger = true;
+                            }
+                        } else if (rec->itemRequired <= 110 && rec->itemRequired - 100 != heldItem) {
+                            // IDs 100 - 110 mean the record will activate when the object is _not_ the specified one
+                            shouldTrigger = true;
+                        }
+                    }
+
+                    if (!shouldTrigger) {
+                        g_nancy->sound->playSound(17); // Hardcoded by original engine
+                    }
+                } else {
+                    shouldTrigger = true;
+                }
+                if (shouldTrigger) {
+                    rec->state = ActionRecord::ExecutionState::kActionTrigger;
+                    
+                    if (rec->itemRequired > 100 && rec->itemRequired <= 110) {
+                        rec->itemRequired -= 100;
+                    }
+
+                    // Re-add the object to the inventory unless it's marked as a one-time use
+                    if (rec->itemRequired == heldItem && rec->itemRequired != -1) {
+                        if (NancySceneState.getInventoryBox().getItemDescription(heldItem).oneTimeUse != 0) {
+                            NancySceneState.getInventoryBox().addItem(heldItem);
+                        }
+
+                        NancySceneState.setHeldItem(-1);
+                    }
+                }
+
+                break;
+            }
+        }
+    }
+}
+
+bool ActionManager::addNewActionRecord(Common::SeekableReadStream &inputData) {
+    inputData.seek(0x30);
+    byte ARType = inputData.readByte();
+    ActionRecord *newRecord = createActionRecord(ARType);
+
+    inputData.seek(0);
+    char *descBuf = new char[0x30];
```

Comment:
This should not get allocated on the heap imo. a `char descBuf[0x30];` should be fine.
(Bonus points for adding a '\0' check here)

Reply:
Fixed.